## Group SQL Practice

_Author: Matt Brems & Katie Sylvia_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [110]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [98]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT * from stores
"""

pd.read_sql_query(sql, engine)

store                         name store_status  \
0      2106  Hillstreet News and Tobacco            A   
1      2112               Mike's Liquors            I   
2      2113               Jamboree Foods            A   
3      2119           Manly Liquor Store            I   
4      2130         Sycamore Convenience            A   
...     ...                          ...          ...   
1968   9917       Honey Creek Distillery            A   
1969   9918  Paradise Distilling Company            A   
1970   9919            Dehner Distillery            A   
1971   9920        Louisiana Spirits LLC            A   
1972   9921                  maDIKwe USA            A   

                                          store_address address_info  
0     2217 College\nCedar Falls, IA 506130000\n(42.5...         None  
1     407 Sharp St.\nGlenwood, IA 515340000\n(41.046...         None  
2     1119 Market St. Box 71\nGowrie, IA 505430000\n...         None  
3     133 East Main\nManly, IA 504560000\n(43.286863...         None  
4     617 Sycamore\nWaterloo, IA 507030000\n(42.4978...         None  
...                                                 ...          ...  
1968  29538, 210th St.\nKeosauqua, IA 52565\n(40.740...         None  
1969  245, Railroad Ave\nDubuque, IA 52001\n(42.4873...         None  
1970  7500, University Ave\nClive, IA\n(41.600352803...         None  
1971  20909, South I-10 Frontage Rd\nLacassine, IA 7...         None  
1972  3101, Ingersoll Ave\nDes Moines, IA 50312\n(41...         None  

[1973 rows x 5 columns]

**Prompt 1 Answer**:

In [59]:
sql = """
SELECT COUNT(store)
FROM stores
"""

pd.read_sql_query(sql, engine)

count
0   1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

**Prompt 2 Answer**:

In [60]:
sql = """
SELECT name
FROM stores
ORDER BY name
LIMIT 4
"""

pd.read_sql_query(sql, engine)

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [61]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, bottle_qty, vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor
LIMIT 10
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty                     vendor
0     $8.99          12  Adamba Imports Int'l Inc.
1    $28.50           6      A Hardy / U.S.A. Ltd.
2    $28.50           1      A Hardy / U.S.A. Ltd.
3    $28.50           3      A Hardy / U.S.A. Ltd.
4    $28.50           6      A Hardy / U.S.A. Ltd.
5    $28.50           6      A Hardy / U.S.A. Ltd.
6    $28.50           6      A Hardy / U.S.A. Ltd.
7    $28.50           3      A Hardy / U.S.A. Ltd.
8    $28.50           3      A Hardy / U.S.A. Ltd.
9    $28.50          30      A Hardy / U.S.A. Ltd.

**Prompt 3 Answer**: 
There are 33 transactions by the same company so the second and third will always be randomized.

1. $8.99  12
2. $28.50  6
3. $28.50  1


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [62]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(total)
FROM sales
WHERE date BETWEEN '2015-02-01' AND '2015-02-28'
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**:

$21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [63]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(vendor) AS num_transactions, SUM(bottle_qty) AS bot_total, SUM(total) AS total_rev, vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY num_transactions
DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

num_transactions  bot_total   total_rev                            vendor
0             42338     403450  7880365.88  Pernod Ricard USA/Austin Nichols

**Prompt 5 Answer**:  Pernod Ricard USA/Austin Nichols with 42,338 transactions with a total of 403,450 bottles sold for a total revenue of $7,880,365.88

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [64]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost AS markup, description, store
FROM sales
ORDER BY markup
DESC
"""

pd.read_sql_query(sql, engine)

markup                           description  store
0        $2,900.00      Cedar Ridge Barrel Proof Bourbon   2588
1        $2,900.00      Cedar Ridge Barrel Proof Bourbon   2590
2          $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3420
3          $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3944
4          $799.60  Absolut w/ Zing Zang Bloody Mary Mix   3354
...            ...                                   ...    ...
3049908      $0.49                Seagrams Extra Dry Gin   3896
3049909      $0.49                Seagrams Extra Dry Gin   3825
3049910      $0.49                Seagrams Extra Dry Gin   3805
3049911      $0.28          Bacardi Black Razz Miniature   2629
3049912      $0.24                     Gran Gala Liqueur   3692

[3049913 rows x 3 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [65]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT ROUND(AVG((btl_price - state_btl_cost)/state_btl_cost)*100) as markup_perc, store
FROM sales
GROUP BY store
ORDER BY markup_perc
DESC
"""

pd.read_sql_query(sql, engine)

markup_perc  store
0            53.0   4013
1            52.0   3833
2            52.0   4266
3            52.0   4024
4            51.0   3713
...           ...    ...
1347         50.0   2595
1348         50.0   2596
1349         50.0   2597
1350         50.0   2599
1351         50.0   2600

[1352 rows x 2 columns]

**Prompt 7 Answer**: 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [66]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.store,
    ROUND(AVG((btl_price - state_btl_cost)/state_btl_cost)*100) as markup_perc,
    ROUND(MIN((btl_price - state_btl_cost)/state_btl_cost)*100) as min_markup,
    ROUND(MAX((btl_price - state_btl_cost)/state_btl_cost)*100) as max_markup,
    stores.name,
    stores.store_address
FROM sales, stores
WHERE sales.store = stores.store
GROUP BY sales.store, stores.name, stores.store_address
ORDER BY markup_perc
DESC
LIMIT 5    
"""

pd.read_sql_query(sql, engine)

store  markup_perc  min_markup  max_markup                        name  \
0   4013         53.0        50.0       671.0      Wal-Mart 0841 / Tipton   
1   4266         52.0        50.0       671.0  Wal-Mart 1683 / Shenandoah   
2   3833         52.0        50.0       671.0    Wal-Mart 3394 / Atlantic   
3   4024         52.0        50.0       671.0  Wal-Mart 1546 / Iowa Falls   
4   2668         51.0        50.0        69.0    Hy-Vee Food Store / Leon   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
2  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
3  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
4  1004 W 1st St\nLeon, IA 50144\n(40.73967779500...

**Prompt 8 Answer**:

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [93]:
sql = """
SELECT counties.county, 
    SUM(bottle_qty)/counties.population AS btl_per_capita,
    ROUND(AVG(products.bottle_size)) AS avg_btl_size
FROM sales, counties, products
WHERE sales.county = counties.county AND products.item_no = sales.item
GROUP BY counties.county
ORDER BY btl_per_capita
DESC
LIMIT 9
"""

pd.read_sql_query(sql, engine)

county  btl_per_capita  avg_btl_size
0      Dickinson              22         974.0
1           Polk              15         893.0
2     Black Hawk              15         863.0
3    Cerro Gordo              14         986.0
4          Scott              13         872.0
5        Johnson              13         895.0
6           Linn              12         886.0
7  Pottawattamie              11         894.0
8        Kossuth              11        1041.0

**Prompt 9 Answer**:

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [116]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, 
    bottle_qty,
    CAST(total AS MONEY)
FROM sales
WHERE total != (btl_price*bottle_qty)
"""

pd.read_sql_query(sql, engine)

ProgrammingError: (psycopg2.errors.UndefinedFunction) operator does not exist: numeric <> money
LINE 6: WHERE total != (btl_price*bottle_qty)
                    ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.

[SQL: 
SELECT btl_price, 
    bottle_qty,
    CAST(total AS MONEY)
FROM sales
WHERE total != (btl_price*bottle_qty)
]
(Background on this error at: https://sqlalche.me/e/20/f405)

**Prompt 10 Answer**: